File: java/src/main/java/topcoder/crazybot/CrazyBot.java
package topcoder.crazybot;

import java.util.Scanner;



public class CrazyBot {
  boolean[][] grid = new boolean[100][100];
  int vx[] = {1, -1, 0, 0};
  int vy[] = {0, 0, 1, -1};

  double[] prob = new double[4];

  public void run() {
    Scanner sc = new Scanner(System.in);
    int n = Integer.parseInt(sc.next());
    int east = Integer.parseInt(sc.next());
    int west = Integer.parseInt(sc.next());
    int south = Integer.parseInt(sc.next());
    int north = Integer.parseInt(sc.next());
    System.out.println(getProbability(n, east, west, south, north));
  }

  double getProbability(int n, int east, int west, int south, int north) {
    prob[0] = east / 100.0;
    prob[1] = west / 100.0;
    prob[2] = south / 100.0;
    prob[3] = north / 100.0;

    return dfs(50, 50, n);
  }

  double dfs(int x, int y, int n) {
    if (grid[x][y]) return 0;
    if (n == 0) return 1;

    grid[x][y] = true;
    double ret = 0;
    for (int i = 0; i < 4; i++ ){
      ret += dfs(x + vx[i], y + vy[i], n - 1) * prob[i];
    }
    grid[x][y] = false; // 別ルートのための初期化
    return ret;
  }
}

File: scala/src/main/scala/topcoder/crazybot/CrazyBot.scala
package topcoder.crazybot

object CrazyBot {

  var k = List()

  def main(args: Array[String]): Unit ={
    val sc = new java.util.Scanner(System.in)
    val n = sc.nextInt
    val east = sc.nextInt
    val weast = sc.nextInt
    val south = sc.nextInt
    val north = sc.nextInt

    val cb = new CrazyBot(east.toDouble/100, weast.toDouble/100, south.toDouble/100, north.toDouble/100)
    println(cb.solve(n, 50, 50))
  }
}

//
class CrazyBot(east: Double, west: Double, south:Double, north: Double) {
  val dirs = List((1, 0, east), (-1, 0, west), (0, 1, south), (0, -1, north))
  val hist = Array.ofDim[Int](100, 100)

  def solve(n: Int, x: Int, y: Int): Double = {
    hist(x)(y) = 1
    val result = n match {
      case 0 => 1
      case _ => {
        var ret = 0.0
        dirs.foreach(d => {
          val nextX = x + d._1
          val nextY = y + d._2
          if(hist(nextX)(nextY) == 0) {
            ret += solve(n - 1, nextX, nextY) * d._3
          }
        })
        ret
      }
    }
    hist(x)(y) = 0
    result
  }
}